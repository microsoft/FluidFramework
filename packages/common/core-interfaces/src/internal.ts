/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

// eslint-disable-next-line no-restricted-syntax
export * from "./index.js";

// Important: all other exports must be type only exports. In package.json exports,
//  index.js is listed as the runtime file. This is done so that all imports are
//  using the same outer runtime file. (Could be changed if needed.)

// Export set of utility types re-tagged as internal for FF client convenience.
// These types are not intended for direct use by customers and api-extractor will
// flag misuse. If an externally visible version of these types is needed, import
// from via /internal/exposedUtilityTypes rather than /internal.
import type { DeepReadonly as ExposedDeepReadonly } from "./deepReadonly.js";
import type { InternalUtilityTypes as ExposedInternalUtilityTypes } from "./exposedInternalUtilityTypes.js";
import type {
	JsonDeserialized as ExposedJsonDeserialized,
	JsonDeserializedOptions,
} from "./jsonDeserialized.js";
import type {
	JsonSerializable as ExposedJsonSerializable,
	JsonSerializableOptions,
} from "./jsonSerializable.js";
import type {
	JsonTypeWith as ExposedJsonTypeWith,
	ReadonlyNonNullJsonObjectWith as ExposedReadonlyNonNullJsonObjectWith,
} from "./jsonType.js";

// Note: There are no docs for these re-exports. `@inheritdoc` cannot be used as:
//   1. api-extractor does not support renames.
//   2. api-extractor does not support package paths. ("Import paths are not supported")
// Also not useful, at least in VS Code, as substitution is not made in place.

/**
 * @internal
 */
export type DeepReadonly<T> = ExposedDeepReadonly<T>;

/**
 * @internal
 */
export type JsonDeserialized<
	T,
	Options extends JsonDeserializedOptions = {
		AllowExactly: [];
		AllowExtensionOf: never;
	},
> = ExposedJsonDeserialized<T, Options>;

/**
 * @internal
 */
export type JsonSerializable<
	T,
	Options extends JsonSerializableOptions = {
		AllowExactly: [];
		AllowExtensionOf: never;
	},
> = ExposedJsonSerializable<T, Options>;

/**
 * @internal
 */
export type JsonTypeWith<T> = ExposedJsonTypeWith<T>;

/**
 * @internal
 */
export type ReadonlyNonNullJsonObjectWith<T> = ExposedReadonlyNonNullJsonObjectWith<T>;

/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/no-namespace
export namespace InternalUtilityTypes {
	/* eslint-disable jsdoc/require-jsdoc */
	export type FlattenIntersection<T extends ExposedInternalUtilityTypes.AnyRecord> =
		ExposedInternalUtilityTypes.FlattenIntersection<T>;
	export type IfSameType<
		X,
		Y,
		IfSame = unknown,
		IfDifferent = never,
	> = ExposedInternalUtilityTypes.IfSameType<X, Y, IfSame, IfDifferent>;
	/* eslint-enable jsdoc/require-jsdoc */
}
