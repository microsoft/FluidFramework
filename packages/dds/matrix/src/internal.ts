/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

// Privately exposed to MatrixUndoProvider
export const undoRemoveRows = Symbol("SharedMatrix.undoRemoveRows");
export const undoRemoveCols = Symbol("SharedMatrix.undoRemoveCols");
