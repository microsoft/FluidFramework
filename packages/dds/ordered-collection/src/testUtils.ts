/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { ConsensusResult, type IConsensusOrderedCollection } from "./interfaces.js";

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 * @internal
 */
export async function acquireAndComplete<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T | undefined> {
	let res: T | undefined;
	await collection.acquire(async (value: T) => {
		res = value;
		return ConsensusResult.Complete;
	});
	return res;
}

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 * @internal
 */
export async function waitAcquireAndComplete<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T> {
	let res: T | undefined;
	await collection.waitAndAcquire(async (value: T) => {
		res = value;
		return ConsensusResult.Complete;
	});
	return res as T;
}

/**
 * Helper method to acquire and release an item
 * Should be used in test code only
 * @internal
 */
export async function acquireAndRelease<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T | undefined> {
	let res: T | undefined;
	await collection.acquire(async (value: T) => {
		res = value;
		return ConsensusResult.Release;
	});
	return res;
}

/**
 * Helper method to acquire and release an item
 * Should be used in test code only
 * @internal
 */
export async function waitAcquireAndRelease<T>(
	collection: IConsensusOrderedCollection<T>,
): Promise<T> {
	let res: T | undefined;
	await collection.waitAndAcquire(async (value: T) => {
		res = value;
		return ConsensusResult.Release;
	});
	return res as T;
}
