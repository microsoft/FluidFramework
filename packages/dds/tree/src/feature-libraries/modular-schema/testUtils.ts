/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { assert } from "@fluidframework/common-utils";
import { Delta } from "../../tree";

export interface MockChildChange {
    intentions: number[];
    ref: number;
}

export function mockChildChangeRebaser(
    change: MockChildChange,
    baseChange: MockChildChange,
): MockChildChange {
    assert(change.ref === baseChange.ref, "Invalid input passed to child rebaser");
    return {
        intentions: change.intentions,
        ref: baseChange.intentions[baseChange.intentions.length - 1],
    };
}

export function mockChildChangeInverter(change: MockChildChange): MockChildChange {
    return {
        intentions: change.intentions.map((i) => -i).reverse(),
        ref: change.intentions[change.intentions.length - 1],
    };
}

export function mockChildChangeComposer(changes: MockChildChange[]): MockChildChange {
    if (changes.length <= 1) {
        return changes[0];
    }
    const id: number[] = [];
    changes.forEach((change, i) => {
        const prev = changes[i - 1].intentions;
        assert(i === 0 || change.ref === prev[prev.length - 1], "Invalid input to child composer");
        id.push(...change.intentions);
    });
    return {
        intentions: id,
        ref: changes[0].ref,
    };
}

export function mockChildChangeToDelta(change: MockChildChange): Delta.Modify {
    return { type: Delta.MarkType.Modify, setValue: { value: change.intentions.map(String).join("|") } };
}
