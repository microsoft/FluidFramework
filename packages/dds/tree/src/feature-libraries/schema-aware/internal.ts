/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

// Used by put public types, but not part of the desired API surface
export {
	TypeSetToTypedTrees as TreeTypesToTypedTreeTypes,
	TypedSchemaData,
	TypedTree,
	CollectOptions,
	TypedFields,
	ApplyMultiplicity,
	ValueFieldTreeFromSchema,
	FlexibleObject,
} from "./schemaAware";

export { NamesFromSchema, ValuesOf, TypedValue, PrimitiveValueSchema } from "./schemaAwareUtil";
