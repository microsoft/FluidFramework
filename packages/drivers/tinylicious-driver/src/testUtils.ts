/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

export { createInsecureTinyliciousTestTokenProvider } from "./insecureTinyliciousTokenProvider.js";

export {
	createInsecureTinyliciousTestUrlResolver,
	createTinyliciousTestCreateNewRequest,
} from "./insecureTinyliciousUrlResolver.js";
