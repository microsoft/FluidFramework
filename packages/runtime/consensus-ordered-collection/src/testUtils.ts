/*!
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */

import {
    ConsensusResult,
    IConsensusOrderedCollection,
} from "./interfaces";

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 */
export async function acquireAndComplete<T>(collection: IConsensusOrderedCollection<T>): Promise<T | undefined> {
    let res: T | undefined;
    await collection.acquire(async (value: T) => {
        res = value;
        return ConsensusResult.Complete;
    });
    return res;
}

/**
 * Helper method to acquire and complete an item
 * Should be used in test code only
 */
export async function waitAcquireAndComplete<T>(collection: IConsensusOrderedCollection<T>): Promise<T> {
    let res: T | undefined;
    await collection.waitAndAcquire(async (value: T) => {
        res = value;
        return ConsensusResult.Complete;
    });
    return res as T;
}
