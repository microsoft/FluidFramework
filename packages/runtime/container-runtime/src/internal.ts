/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

/* Exports for internal use only
 * This is only used during transition of legacy+alpha code to internal.
 * Here there are conflicts between index reexports with deprecations and clean internal versions.
 * In cases where * exports conflict with named exports, the named exports take precedence per
 * https://tc39.es/ecma262/multipage/ecmascript-language-scripts-and-modules.html#sec-getexportednames.
 * This does trigger the `import/export` lint warning (which is intentionally disabled here). This
 * approach ensures that a non-deprecated version of ContainerRuntime (provided as named direct
 * export) eclipses the deprecated one from `./index.ts`.
 * Alternatively, another file could be created for external exports adding ContainerRuntime and
 * they'd share common exports files (the rest of current index.ts).
 */
/* eslint-disable import/export */
/* eslint-disable import/no-deprecated */

// eslint-disable-next-line no-restricted-syntax
export * from "./index.js";
import { ContainerRuntime as ContainerRuntimeClass } from "./containerRuntime.js";
export type ContainerRuntime = ContainerRuntimeClass;
export const ContainerRuntime = ContainerRuntimeClass;
