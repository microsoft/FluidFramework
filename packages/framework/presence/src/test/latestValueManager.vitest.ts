/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { EventAndErrorTrackingLogger } from "@fluidframework/test-utils/internal";
import { useFakeTimers, type SinonFakeTimers } from "sinon";
import {
	describe,
	it,
	afterAll,
	afterEach,
	beforeAll,
	beforeEach,
} from "vitest";

import { Latest, Notifications, type PresenceNotifications } from "../index.js";
import type { createPresenceManager } from "../presenceManager.js";

import { MockRuntimeSignalSnapshotter } from "./snapshotEphemeralRuntime.js";
import { generateBasicClientJoin, prepareConnectedPresence } from "./testUtils.js";

describe("Presence", () => {
	describe("LatestValueManager", () => {
		describe("batching", () => {
			// IMPORTANT: All tests in this suite have an extra signal for each workspace that is initialized in the test.
			// This is a bug. See AB#24392. This means that when looking at snapshots from these tests, the "real" snapshots
			// start at 2, not 1, for most tests. Some tests may have additional incorrect leading signals. Such examples
			// are noted inline.
			let runtime: MockRuntimeSignalSnapshotter;
			let logger: EventAndErrorTrackingLogger;
			const initialTime = 1000;
			let clock: SinonFakeTimers;
			let presence: ReturnType<typeof createPresenceManager>;

			beforeAll(async () => {
				clock = useFakeTimers();
			});

			beforeEach(() => {
				logger = new EventAndErrorTrackingLogger();
				runtime = new MockRuntimeSignalSnapshotter(logger);

				// We are configuring the runtime to be in a connected state, so ensure it looks connected
				runtime.connected = true;

				// Note that while the initialTime is set to 1000, the prepareConnectedPresence call advances
				// it to 1010 so all tests start at that time.
				clock.setSystemTime(initialTime);

				// Disable submitting signals with a dummy function. This ensures we don't capture signals from
				// test setup, like the prepareConnectedPresence call.
				const submitSignalOriginal = runtime.submitSignal;
				runtime.submitSignal = () => {};

				// Set up the presence connection
				presence = prepareConnectedPresence(runtime, "sessionId-2", "client2", clock, logger);

				// Restore the submiSignal function
				runtime.submitSignal = submitSignalOriginal;
			});

			afterEach(() => {
				clock.reset();
			});

			afterAll(() => {
				clock.restore();
			});

			it("sends signal immediately when allowable latency is 0", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatency: 0, forcedRefreshInterval: 0 }),
				}); // SIGNAL #1 DUE TO AB#24392

				const { count } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020

				// SIGNAL #2
				count.local = { num: 42 };
			});

			it("batches signals sent within the allowableUpdateLatency", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatency: 100, forcedRefreshInterval: 0 }),
				}); // SIGNAL #1 DUE TO AB#24392

				const { count } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020
				count.local = { num: 12 }; // will be queued; deadline is set to 1120

				clock.tick(80); // Time is now 1100
				count.local = { num: 34 }; // will be queued; deadline remains 1120

				clock.tick(30); // Time is now 1130
				const expectedClientJoin = generateBasicClientJoin(clock.now, {
					clientSessionId: "sessionId-3",
					clientConnectionId: "client3",
					updateProviders: ["client2"],
					averageLatency: 10,
				});
				presence.processSignal("", expectedClientJoin, true);
				// SIGNAL #2
				// The deadline has now passed, so the timer will fire and send a single
				// signal with the value from the last signal (num=34).

				clock.tick(10); // Time is now 1140
				count.local = { num: 56 }; // will be queued; deadline is set to 1240

				clock.tick(40); // Time is now 1180
				count.local = { num: 78 }; // will be queued; deadline remains 1240

				clock.tick(40); // Time is now 1220
				count.local = { num: 90 }; // will be queued; deadline remains 1240

				clock.tick(30); // Time is now 1250
				// SIGNAL #3
				// The deadline has now passed, so the timer will fire and send a single
				// signal with the value from the last signal (num=90).
			});

			it("queued signal is sent immediately with immediate update message", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatency: 100, forcedRefreshInterval: 0 }),
					immediateUpdate: Latest(
						{ num: 0 },
						{ allowableUpdateLatency: 0, forcedRefreshInterval: 0 },
					),
				}); // SIGNAL #1 DUE TO AB#24392

				const { count, immediateUpdate } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020
				count.local = { num: 12 }; // will be queued; deadline is set to 1120

				clock.tick(80); // Time is now 1100
				count.local = { num: 34 }; // will be queued; deadline remains 1120

				clock.tick(10); // Time is now 1110
				immediateUpdate.local = { num: 56 };
				// SIGNAL #2
				// This should cause the queued signals to be merged with this immediately-sent
				// signal with the value from the last signal (num=34).
			});

			it("batches signals with different allowed latencies", async () => {
				// Configure a state workspace
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatency: 100, forcedRefreshInterval: 0 }),
					note: Latest(
						{ message: "" },
						{ allowableUpdateLatency: 50, forcedRefreshInterval: 0 },
					),
				}); // SIGNAL #1 DUE TO AB#24392

				const { count, note } = stateWorkspace.props;

				clock.tick(10); // Time is now 1020
				note.local = { message: "will be queued" }; // will be queued, deadline is set to 1070
				count.local = { num: 12 }; // will be queued; deadline remains 1070

				clock.tick(30); // Time is now 1050
				count.local = { num: 34 }; // will be queued; deadline remains 1070

				clock.tick(10); // Time is now 1060
				note.local = { message: "final message" }; // will be queued; deadline remains 1070

				clock.tick(30); // Time is now 1080
				// SIGNAL #2
				// The deadline has now passed, so the timer will fire and send a single
				// signal with the value from the last signal (num=34, message="final message").
			});

			it("batches signals from multiple workspaces", async () => {
				// Configure two state workspaces
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatency: 100, forcedRefreshInterval: 0 }),
				}); // SIGNAL #1 DUE TO AB#24392
				const stateWorkspace2 = presence.getStates("name:testStateWorkspace2", {
					note: Latest(
						{ message: "" },
						{ allowableUpdateLatency: 50, forcedRefreshInterval: 0 },
					),
				}); // SIGNAL #2 DUE TO AB#24392

				const { count } = stateWorkspace.props;
				const { note } = stateWorkspace2.props;

				clock.tick(10); // Time is now 1020
				note.local = { message: "will be queued" }; // will be queued, deadline is set to 1070
				count.local = { num: 12 }; // will be queued; deadline remains 1070

				clock.tick(30); // Time is now 1050
				count.local = { num: 34 }; // will be queued; deadline remains 1070

				clock.tick(10); // Time is now 1060
				note.local = { message: "final message" }; // will be queued; deadline remains 1070

				// Messages are auto-sent at time 1070, the deadline

				clock.tick(30); // Time is now 1090
				// SIGNAL #3
				// The deadline has now passed, so the timer will fire at time 1070 and send a single
				// signal with the values from the last workspace updates (num=34, message="final message").
			});

			it("notification signals are sent immediately", async () => {
				// Configure a notifications workspaces
				// eslint-disable-next-line @typescript-eslint/ban-types
				const notificationsWorkspace: PresenceNotifications<{}> = presence.getNotifications(
					"name:testNotificationWorkspace",
					{},
				);

				notificationsWorkspace.add(
					"testEvents",
					Notifications<
						// Below explicit generic specification should not be required.
						{
							newId: (id: number) => void;
						},
						"testEvents"
					>(
						// A default handler is not required
						{
							newId: (client, newId) => {
								// do nothing?
							},
						},
					),
				);

				const { testEvents } = notificationsWorkspace.props;

				clock.tick(10); // Time is now 1020

				clock.tick(30); // Time is now 1050
				testEvents.emit.broadcast("newId", 77);
				// SIGNAL #1

				clock.tick(10);
				testEvents.emit.broadcast("newId", 99);
				// SIGNAL #2
			});

			// TODO: RESULTS NOT VALID!!!
			it("notification signals cause queued messages to be sent immediately", async () => {
				// Configure a state workspaces
				const stateWorkspace = presence.getStates("name:testStateWorkspace", {
					count: Latest({ num: 0 }, { allowableUpdateLatency: 100, forcedRefreshInterval: 0 }),
				}); // SIGNAL #1 DUE TO AB#24392

				// eslint-disable-next-line @typescript-eslint/ban-types
				const notificationsWorkspace: PresenceNotifications<{}> = presence.getNotifications(
					"name:testNotificationWorkspace",
					{},
				);

				notificationsWorkspace.add(
					"testEvents",
					Notifications<
						// Below explicit generic specification should not be required.
						{
							newId: (id: number) => void;
						},
						"testEvents"
					>(
						// A default handler is not required
						{},
					),
				);

				const { count } = stateWorkspace.props;
				const { testEvents } = notificationsWorkspace.props;

				testEvents.notifications.on("newId", (client, newId) => {
					// do nothing
				});

				clock.tick(10); // Time is now 1020
				count.local = { num: 12 }; // will be queued, deadline is set to 1120

				clock.tick(30); // Time is now 1050
				testEvents.emit.broadcast("newId", 99);
				// SIGNAL #2
				// The deadline has now passed, so the timer will fire and send a
				// signal with the value from the last signal (num=12)
				// There should also be a signal for the notification, which is NOT
				// being sent
			});
		});
	});
});
