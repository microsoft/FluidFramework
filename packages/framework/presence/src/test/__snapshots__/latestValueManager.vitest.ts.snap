// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Presence > LatestValueManager > batching > batches signals sent within the allowableUpdateLatency > submitted signal 1`] = `
"[
  "Pres:ClientJoin",
  {
    "sendTimestamp": 1000,
    "avgLatency": 0,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      }
    },
    "updateProviders": [
      "client0",
      "client1",
      "client3"
    ]
  }
]"
`;

exports[`Presence > LatestValueManager > batching > batches signals sent within the allowableUpdateLatency > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1010,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "data": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "num": 0
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > LatestValueManager > batching > batches signals sent within the allowableUpdateLatency > submitted signal 3`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1100,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "data": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1100,
            "value": {
              "num": 34
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > LatestValueManager > batching > batches signals sent within the allowableUpdateLatency > submitted signal 4`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1220,
    "avgLatency": 5,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "data": {
          "sessionId-2": {
            "rev": 5,
            "timestamp": 1220,
            "value": {
              "num": 90
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > LatestValueManager > batching > batches signals with different allowed latencies > submitted signal 1`] = `
"[
  "Pres:ClientJoin",
  {
    "sendTimestamp": 1000,
    "avgLatency": 0,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      }
    },
    "updateProviders": [
      "client0",
      "client1",
      "client3"
    ]
  }
]"
`;

exports[`Presence > LatestValueManager > batching > batches signals with different allowed latencies > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1010,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "data": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "num": 0
            }
          }
        },
        "otherData": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "message": ""
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > LatestValueManager > batching > batches signals with different allowed latencies > submitted signal 3`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1060,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "otherData": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1060,
            "value": {
              "message": "final message"
            }
          }
        },
        "data": {
          "sessionId-2": {
            "rev": 2,
            "timestamp": 1050,
            "value": {
              "num": 34
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > LatestValueManager > batching > sends signal immediately when allowable latency is 0 > submitted signal 1`] = `
"[
  "Pres:ClientJoin",
  {
    "sendTimestamp": 1000,
    "avgLatency": 0,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      }
    },
    "updateProviders": [
      "client0",
      "client1",
      "client3"
    ]
  }
]"
`;

exports[`Presence > LatestValueManager > batching > sends signal immediately when allowable latency is 0 > submitted signal 2`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1010,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "data": {
          "sessionId-2": {
            "rev": 0,
            "timestamp": 1010,
            "value": {
              "num": 0
            }
          }
        }
      }
    }
  }
]"
`;

exports[`Presence > LatestValueManager > batching > sends signal immediately when allowable latency is 0 > submitted signal 3`] = `
"[
  "Pres:DatastoreUpdate",
  {
    "sendTimestamp": 1020,
    "avgLatency": 10,
    "data": {
      "system:presence": {
        "clientToSessionId": {
          "client2": {
            "rev": 0,
            "timestamp": 1000,
            "value": "sessionId-2"
          }
        }
      },
      "s:name:testStateWorkspace": {
        "data": {
          "sessionId-2": {
            "rev": 1,
            "timestamp": 1020,
            "value": {
              "num": 42
            }
          }
        }
      }
    }
  }
]"
`;
